cmd/syncComparator.go

```diff
@@ -54,30 +202,217 @@ func newSyncDestinationComparator(i *folderIndexer, copyScheduler, cleaner objec
+	var present bool
+	var sourceObjectInMap StoredObject
+
+	//
+	// We come here with 'destinationObject' when Target traverser (parallelList() for Target=Blob) enumerates a directory from tqueue and
+	// creates and enqueues a destinationObject for each enumerated children. Target traverser also creates and enqueues a special
+	// destinationObject (with isFolderEndMarker==true) after queueing destinationObjects for all direct children of the directory.
+	// This special destinationObject is guaranteed to be queued *after* all the other children objects for the directory (if any).
+	// See below for more details on the special isFolderEndMarker object.
+	//
+	if f.sourceFolderIndex.isDestinationCaseInsensitive {
+		lcRelativePath = strings.ToLower(destinationObject.relativePath)
+	} else {
+		lcRelativePath = destinationObject.relativePath
 	}
 
-	// if the destinationObject is present at source and stale, we transfer the up-to-date version from source
-	if present {
-		defer delete(f.sourceIndex.indexMap, destinationObject.relativePath)
-		if f.disableComparison || sourceObjectInMap.isMoreRecentThan(destinationObject) {
-			err := f.copyTransferScheduler(sourceObjectInMap)
-			if err != nil {
-				return err
+	//
+	// parent dir name, will be "." for files/dirs at the copy root.
+	//
+	// So, for lcRelativePath == "dir1/dir2",
+	// lcFolderName -> "dir1"
+	// lcFileName   -> "dir2"
+	//
+	// and for "dir1" (dir1 present at copy root),
+	// lcFolderName -> "."
+	// lcFileName   -> "dir1"
+	//
+	// Q: When will folderPresent be false below?
+	// A: Since we process every directory independently, if "dir1" is fully processed before "dir1/dir2",
+	//    when processIfNecessary() is called for "dir1/dir2" with isFolderEndMarker==true, foldermap["dir1"]
+	//    won't be found as it would have been deleted after "dir1" was fully processed, though we MUST have
+	//    foldermap["dir1/dir2"].
+	//    Also note that folderPresent MUST never be false when isFolderEndMarker==false, since target traverser
+	//    only ever enumerates directories added to tqueue by the source traverser and all such directories
+	//    MUST be added to folderMap by the source traverser.
+	//
+	lcFolderName = filepath.Dir(lcRelativePath)
+	lcFileName = filepath.Base(lcRelativePath)
+
+	f.sourceFolderIndex.lock.Lock()
+
+	foldermap, folderPresent := f.sourceFolderIndex.folderMap[lcFolderName]
+
+	// Do the auditing of map.
+	defer func() {
+		if folderPresent {
+			if len(f.sourceFolderIndex.folderMap[lcFolderName].indexMap) == 0 {
+				size := int64(unsafe.Sizeof(objectIndexer{}))
+				atomic.AddInt64(&f.sourceFolderIndex.totalSize, -size)
+				delete(f.sourceFolderIndex.folderMap, lcFolderName)
 			}
 		}
-	} else {
-		// purposefully ignore the error from destinationCleaner
-		// it's a tolerable error, since it just means some extra destination object might hang around a bit longer
-		_ = f.destinationCleaner(destinationObject)
+		f.sourceFolderIndex.lock.Unlock()
+	}()
```

Comment:
as discussed let's delete folderMap[folder] from FinalizeTargetDirectory. This is not needed then.